app/helpers/ExerciseConfig/Compilation/BoxesOptimizer.php

```diff
@@ -14,19 +17,255 @@
+    foreach ($removedNode->getBox()->getOutputPorts() as $outputPort) {
+      $variable = $outputPort->getVariableValue();
+      if ($variable) {
+        $oldVars[] = $variable;
+        $newVars[] = $node->getBox()->getOutputPort($outputPort->getName())->getVariableValue();
+      }
+    }
+
+    // repair variables in all children of removed node
+    foreach ($removedNode->getPortNode()->getChildren() as $child) {
+      foreach ($child->getBox()->getInputPorts() as $inputPort) {
+        $searchIt = array_search($inputPort->getVariableValue(), $oldVars, true);
+        if ($searchIt !== false) {
+          $inputPort->setVariableValue($newVars[$searchIt]);
+        }
+      }
+    }
+  }
+
+  /**
+   * In-place optimisation of the given tree.
+   * @param Node $rootNode
+   */
+  private function optimizeTree(Node $rootNode) {
+    // Ok, here it goes...
+    // The whole optimisation is based on following heuristic. We were given a
+    // rooted tree which can have multiple root nodes. The tree is traversed by
+    // levels and can be implemented with recursion. At first all root nodes are
+    // compared if there are any duplicates. If duplicates are found, then these
+    // nodes are merged into one. The next step is to process all subtrees which
+    // were created. Thus if there were 4 nodes and 2 and 2 are the same, these
+    // four nodes are contracted into 2 nodes. These two nodes then contain
+    // subtrees from the 2 nodes of which they are composed. After this,
+    // the procedure is repeated for all nodes from subtrees.
+    // Therefore this heuristics is capable only optimise the beginning of the
+    // trees and not the ends. This is generally fine for our usage, because
+    // usually the same thing for all tests is compilation which is the first
+    // set of tasks in the tree.
+
+    // processing queue simulating breadth-first search
+    $queue = [$rootNode];
+    while (!empty($queue)) {
+      $node = array_shift($queue); /** @var Node $node */
+      $children = $node->getChildren();
+
+      if (count($children) < 2) {
+        // there is nothing to optimize if node has only one or none children
+        continue;
+      }
+
+      // go through all children from currently processed node
+      // all children are taken as a base for further comparisons
+      while (!empty($children)) {
+        $child = array_shift($children);
+        $copyChildren = $children;
+
+        // base $child has to be compared against all other children
+        // if the children are equal, then they should be optimised
+        while(!empty($copyChildren)) {
+          $compared = array_shift($copyChildren);
+
+          // children can be optimized... so do it
+          if (self::canNodesBeOptimized($child, $compared)) {
+            // delete the children from the base children array
+            $children = array_filter($children, function ($n) use ($compared) {
+              return $n === $compared;
```

Comment:
I am not sure, but I think that this condition should be reversed.

Reply:
It looks like, that it should...